WhateverGreen/kern_igfx_clock.cpp

```diff
@@ -0,0 +1,881 @@
+		DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Will use the maximum link rate specified by user or cached by the previous probe call.");
+		caps->maxLinkRate = callbackIGFX->modDPCDMaxLinkRateFix.maxLinkRate;
+	} else {
+		DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Will probe the maximum link rate from the table.");
+		caps->maxLinkRate = callbackIGFX->modDPCDMaxLinkRateFix.probeMaxLinkRate();
+		// The graphics driver tries to read the maximum link rate from both DPCD address 0x0 and 0x2200.
+		// We save the probe result to avoid duplicated computation.
+		callbackIGFX->modDPCDMaxLinkRateFix.maxLinkRate = caps->maxLinkRate;
+	}
+	
+	// All done
+	DBGLOG("igfx", "MLR: [COMM] wrapReadAUX() Maximum link rate 0x%02x has been set in the DPCD buffer.", caps->maxLinkRate);
+	return retVal;
+}
+
+IOReturn IGFX::DPCDMaxLinkRateFix::orgReadAUX(uint32_t address, void *buffer, uint32_t length) {
+	if (port != nullptr) {
+		// ICL+
+		DBGLOG("igfx", "MLR: [COMM] orgReadAUX() Routed to ICL IMP with Address = 0x%x; Length = %u.", address, length);
+		return orgICLReadAUX(port, address, buffer, length);
+	} else {
+		// CFL-
+		DBGLOG("igfx", "MLR: [COMM] orgReadAUX() Routed to CFL IMP with Address = 0x%x; Length = %u.", address, length);
+		return orgCFLReadAUX(controller, framebuffer, address, length, buffer, displayPath);
+	}
+}
+
+bool IGFX::DPCDMaxLinkRateFix::getFramebufferIndex(uint32_t &index) {
+	auto framebuffer = port != nullptr ? orgICLGetFBFromPort(*callbackIGFX->gFramebufferController, port) : this->framebuffer;
+	DBGLOG("igfx", "MLR: [COMM] GetFBIndex() Port at 0x%llx; Framebuffer at 0x%llx.", port, framebuffer);
+	return AppleIntelFramebufferExplorer::getIndex(framebuffer, index);
+}
+
+uint32_t IGFX::DPCDMaxLinkRateFix::probeMaxLinkRate() {
+	// Precondition: This function is only called when the framebuffer index is 0 (i.e. builtin display)
+	// Guard: Read the eDP version from DPCD
+	uint8_t eDPVersion;
+	if (orgReadAUX(DPCD_EDP_VERSION_ADDRESS, &eDPVersion, 1) != kIOReturnSuccess) {
+		SYSLOG("igfx", "MLR: [COMM] ProbeMaxLinkRate() Failed to read the eDP version. Aborted.");
+		return 0;
+	}
+	
+	// Guard: Ensure that eDP is >= 1.4
+	if (eDPVersion < DPCD_EDP_VERSION_1_4_VALUE) {
+		SYSLOG("igfx", "MLR: [COMM] ProbeMaxLinkRate() eDP version is less than 1.4. Aborted.");
+		return 0;
+	}
+	DBGLOG("igfx", "MLR: [COMM] ProbeMaxLinkRate() Found eDP version 1.4+ (Value = 0x%x).", eDPVersion);
+	
+	// Guard: Read all supported link rates
+	uint16_t rates[DP_MAX_NUM_SUPPORTED_RATES] = {0};
+	if (orgReadAUX(DPCD_EDP_SUPPORTED_LINK_RATES_ADDRESS, rates, sizeof(rates)) != kIOReturnSuccess) {
+		SYSLOG("igfx", "MLR: [COMM] ProbeMaxLinkRate() Failed to read supported link rates from DPCD.");
+		return 0;
+	}
+	
+	// Parse all supported link rates reported by DPCD
+	// The last non-zero entry in the table is the maximum link rate supported by the eDP 1.4 panel
+	uint32_t last = 0;
+	for (int index = 0; index < arrsize(rates); index += 1) {
```

Comment:
Maybe `index++`?

More importantly, is it guaranteed that the table is sorted? In the description you said we choose the highest value.

Reply:
Based on my observation so far, the table is sorted in ascending order, so the last non-zero element should be the maximum link rate supported by the panel.

I have added a guard to ensure that we will get the largest value from the table, in case that a "hateful" panel does not conform to the standard.